### PR TITLE
Recurse into subgraphs in transformers and session initialization

### PR DESCRIPTION
**Pull request by skottmckay**
The old session initialization code processed the main graph first and subgraphs later. Transformers were inconsistent and most were not aware of the need to recurse into a subgraph so that any transformations are also applied there. Graph::Resolve was being called for every level of the graph leading to a lot of unnecessary processing part way through the transform (Graph::Resolve always goes up to the main gain and descends through all subgraphs, so calling it each time an individual subgraph was changed was generally unnecessary).

There are also use cases where we need to have subgraphs fully processed as early as possible in order to make optimal decisions. For example, we need the subgraph to have all transforms applied and be fully partitioned in order to be able to avoid a copy to/from CPU to run a control flow node if the subgraph is going to run on GPU, and the input data is on GPU and not needed directly by the control flow node. Without the partitioning we don't know where the subgraph will run and incur an unnecessary copy to/from CPU based on the control flow node always running there.

Change GraphTransformer to
  - call Graph::Resolve in Apply before and after transform at the top level only for consistency
  - move the Apply method to a private virtual method called ApplyImpl and override that in the derived classes 
    - ensures GraphTransformer::Apply runs the transform
    - add a graph_level parameter so the implementation knows if it's processing the main graph or a subgraph
  - add a Recurse method to make it easy to process any subgraphs (Scan/If/Loop operator) in a node in the graph

Update existing GraphTransformer classes to new API and to call Recurse.
Remove unnecessary calls to Graph::Resolve in the transformers.

Update inference session initialization to recurse into subgraphs
  - separate creation of SessionState for subgraphs which is needed by some transformers from the creation of the execution plan etc. for the subgraph

Split out second phase of InsertCastTransformer where duplicates are removed into a separate GraphTransformer so we can run that after the first phase completes inserting Cast nodes.

### CODE REVIEW

**Comment by linkerzhang** [comment]
is it possible that a control flow node in main graph may be updated (changed or removed) during graph transformation please? if yes, I think the current recursive transformation for subgraphs may not be correct. The workflow in my mind should be, optimize the main graph (with all transformers) -> optimize sub graphs (with all transformers). 

The logic in this PR is, Run all transformers (for main graph and then sub graphs).

**Comment by linkerzhang** [comment]
Is graph resolve the perf hit? if yes, graph transformers should be updated to not call any graph resolve, given we have node/edge editing APIs in place. That means, graph transformer should be responsible to make a graph consistent/valid after its running (they depend on resolve to ensure this before).

File: onnxruntime/core/session/inference_session.cc [hunk 1]
```diff
@@ -409,25 +407,20 @@ class InferenceSession::Impl {
       SessionStateInitializer session_initializer{graph, session_state_, execution_providers_,
                                                   kernel_registry_manager_};
 
+      // create SessionState for subgraphs as it's needed by the transformers
+      ORT_RETURN_IF_ERROR(CreateSubgraphSessionState(graph, session_state_));
```
Comment (weixingzhang) [inline onnxruntime/core/session/inference_session.cc]: I am creating a separate project named optimizer and moving all graph optimization related code to this new project. In the meantime, I am thinking to remove dependency on session_state from graph optimization.

**Comment by skottmckay** [comment]
The new structure lets the transformer decide when to process the subgraphs as I'm not sure there's one single answer for what the best order is. The transformer can do the main graph first and subgraphs after that, or it can recurse into subgraphs as it processes the control flow node containing the subgraph. The user controls the order that the transformers will run in, and each registered transformer will be run once for the model.

Is there a scenario where this approach is a problem?  

Another issue with doing the main graph and subgraphs separately is that you have to call Graph::Resolve between each transformer that changes the graph. Graph::Resolve operates on the entire graph and not just a specific subgraph as it has to build the connections for implicit inputs etc. starting at the topmost graph, so you potentially get (num transformers x (main graph + num subgraphs)) calls to Graph::Resolve, with Graph::Resolve processing (main graph + num subgraphs) Graph instances every time.

---
In reply to: [457785483](https://github.com/Microsoft/onnxruntime/pull/368#issuecomment-457785483) [](ancestors = 457785483)

**Comment by skottmckay** [comment]
They still use Graph::Resolve the to ensure the graph is valid if it's modified as before, but instead of doing that in lots of places it just does it once before the transformer runs and once after the transformer fully completes (if the modified flag is set) by default. 

---
In reply to: [457786565](https://github.com/Microsoft/onnxruntime/pull/368#issuecomment-457786565) [](ancestors = 457786565)

File: onnxruntime/core/framework/graph_partitioner.h [hunk 2]
```diff
@@ -17,12 +17,11 @@ class SessionState;
...
  public:
   //The order of providers represents the user preference.
-  GraphPartitioner(KernelRegistryManager& kernel_registry_mgr,
-                   const ExecutionProviders& providers)
+  GraphPartitioner(KernelRegistryManager& kernel_registry_mgr, const ExecutionProviders& providers)
       : kernel_registry_mgr_(kernel_registry_mgr),
         providers_(providers) {}
 
-  Status Partition(onnxruntime::Graph& graph, bool export_dll, FuncManager* func_mgr) const;
+  Status Partition(Graph& graph, SessionState& session_state) const;
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/framework/graph_partitioner.h]: The partitioning needs to get the subgraph session states, so I could pass the single SessionState to provides access to those, or 3 parameters (export_dll, funcmgr and the map for the subgraph session states). Given that, it seemed to have reached a point where it made more sense to just pass the session state.

---
In reply to: [251174006](https://github.com/Microsoft/onnxruntime/pull/368#discussion_r251174006) [](ancestors = 251174006)

File: onnxruntime/core/session/inference_session.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/session/inference_session.cc]: Would be great if that can be done.

---
In reply to: [251263710](https://github.com/Microsoft/onnxruntime/pull/368#discussion_r251263710) [](ancestors = 251263710)

**Comment by gramalingam** [comment]
The generic strategy sounds reasonable to me. Agree that there is no single answer for the best order in all cases.

**Pull request merged**